sites/avivator/src/state.js

```diff
@@ -129,27 +132,65 @@ const DEFAULT_VIEWER_STATE = {
   pyramidResolution: 0
 };
 
-export const useViewerStore = create(set => ({
-  ...DEFAULT_VIEWER_STATE,
-  ...generateToggles(DEFAULT_VIEWER_STATE, set),
-  setIsChannelLoading: (index, val) =>
-    set(state => {
-      const newIsChannelLoading = [...state.isChannelLoading];
-      newIsChannelLoading[index] = val;
-      return { ...state, isChannelLoading: newIsChannelLoading };
-    }),
-  addIsChannelLoading: val =>
-    set(state => {
-      const newIsChannelLoading = [...state.isChannelLoading, val];
-      return { ...state, isChannelLoading: newIsChannelLoading };
-    }),
-  removeIsChannelLoading: index =>
-    set(state => {
-      const newIsChannelLoading = [...state.isChannelLoading];
-      newIsChannelLoading.splice(index, 1);
-      return { ...state, isChannelLoading: newIsChannelLoading };
-    })
-}));
+const createViewerStore = () =>
+  createStore(set => ({
+    ...DEFAULT_VIEWER_STATE,
+    ...generateToggles(DEFAULT_VIEWER_STATE, set),
+    setIsChannelLoading: (index, val) =>
+      set(state => {
+        const newIsChannelLoading = [...state.isChannelLoading];
+        newIsChannelLoading[index] = val;
+        return { ...state, isChannelLoading: newIsChannelLoading };
+      }),
+    addIsChannelLoading: val =>
+      set(state => {
+        const newIsChannelLoading = [...state.isChannelLoading, val];
+        return { ...state, isChannelLoading: newIsChannelLoading };
+      }),
+    removeIsChannelLoading: index =>
+      set(state => {
+        const newIsChannelLoading = [...state.isChannelLoading];
+        newIsChannelLoading.splice(index, 1);
+        return { ...state, isChannelLoading: newIsChannelLoading };
+      })
+  }));
+
+export const AvivatorContext = createContext(null);
+
+export const AvivatorProvider = ({ children }) => {
+  const storesRef = useRef(null);
+  if (!storesRef.current) {
+    storesRef.current = {
+      channels: createChannelsStore(),
+      imageSettings: createImageSettingsStore(),
+      viewer: createViewerStore()
+    };
+  }
+  return createElement(
+    AvivatorContext.Provider,
+    { value: storesRef.current },
+    children
+  );
+};
+
+function useStoreApi(storeName) {
+  const store = useContext(AvivatorContext);
+  if (!store) throw 'useStore must be used within a AvivatorProvider';
+  return store[storeName];
+}
+export const useChannelsStoreApi = () => useStoreApi('channels');
+export const useImageSettingsStoreApi = () => useStoreApi('imageSettings');
+export const useViewerStoreApi = () => useStoreApi('viewer');
+function useAviStore(storeName, selector, eqFn) {
+  const store = useStoreApi(storeName);
+  return useStoreWithEqualityFn(store, selector, eqFn);
+}
+export const useChannelsStore = (selector, eqFn) =>
+  useAviStore('channels', selector, eqFn);
+export const useImageSettingsStore = (selector, eqFn) =>
+  useAviStore('imageSettings', selector, eqFn);
+export const useViewerStore = (selector, eqFn) =>
+  useAviStore('viewer', selector, eqFn);
```

Comment:
> Can you explain the "architecture," maybe just using docstrings? 

Will do.

Reply:
@ilan-gold I've added some explanation, lmk if it is clear etc.

It's just occurred to me that there is probably another change needed to assign unique IDs to views - should be reasonably straightforward, although would also be quite easy to miss something - with a risk of that leaving potential obscure bugs lurking under the surface...

I think any such bugs would only apply in instances where multiple contexts exist, but you'd be very justified in seeing it as an extra potential red-flag: please don't feel you're committed to accepting these changes. I'm going to do a bit more work on the MDV side - hopefully get that into a better state than the one I'd left it in recently 🙄, probably write/copy a bit more code that I hope to delete, and get more of a sense of how useful it'd be to have more of avivator available to import...

Reply:
FWIW in case interested, [this code-sandbox](//https://codesandbox.io/p/devbox/zustand-context-ts-y5cpyh) is a minimal version of the same pattern with TypeScript types.